File: scripts/validate-data.js
#!/usr/bin/env node
/**
 * Validate the personal-project datasets under /data.
 * Run from the repo root: node scripts/validate-data.js
 */
"use strict";

const fs = require("fs");
const path = require("path");

const root = path.join(__dirname, "..", "data");
const catalog = JSON.parse(fs.readFileSync(path.join(root, "catalog.json"), "utf8"));

const errors = [];
const info = [];

function fail(msg) {
  errors.push(msg);
}

function ok(msg) {
  info.push(msg);
}

function readJson(file) {
  const full = path.join(root, file);
  if (!fs.existsSync(full)) {
    fail("missing file " + file);
    return null;
  }
  try {
    return JSON.parse(fs.readFileSync(full, "utf8"));
  } catch (err) {
    fail(file + " is not valid JSON: " + err.message);
    return null;
  }
}

function isNumberArray(arr) {
  return Array.isArray(arr) && arr.every(function (n) {
    return typeof n === "number" && Number.isFinite(n);
  });
}

function sameLength() {
  const arrays = Array.prototype.slice.call(arguments);
  const first = arrays[0];
  return arrays.every(function (arr) {
    return arr.length === first.length;
  });
}

catalog.datasets.forEach(function (entry) {
  const data = readJson(entry.file);
  if (!data) {
    return;
  }
  if (data.id !== entry.id) {
    fail(entry.file + " id mismatch: " + data.id + " !== " + entry.id);
  } else {
    ok(entry.file + " id matches catalog");
  }
});

const heightShoe = readJson("height-shoe-size.json");
if (heightShoe) {
  ["female", "male"].forEach(function (sex) {
    const series = heightShoe.series[sex];
    const n = series.points.length;
    const fit = series.fit;
    if (n !== fit.n) {
      fail("height-shoe-size " + sex + " fit.n " + fit.n + " !== points " + n);
    }
    series.points.forEach(function (p, i) {
      if (typeof p.height !== "number" || typeof p.shoeLength !== "number") {
        fail("height-shoe-size " + sex + " point " + i + " missing numbers");
      }
    });
    if (fit.r2 < 0.7) {
      fail("height-shoe-size " + sex + " unexpected weak fit r2=" + fit.r2);
    }
    ok("height-shoe-size " + sex + " n=" + n + " r2=" + fit.r2);
  });
}

const ageFoot = readJson("age-foot-length.json");
if (ageFoot) {
  if (!sameLength(ageFoot.ages, ageFoot.series.female, ageFoot.series.male, ageFoot.differenceFemaleMinusMale)) {
    fail("age-foot-length series lengths differ");
  } else {
    ok("age-foot-length aligned length=" + ageFoot.ages.length);
  }
  ageFoot.ages.forEach(function (age, i) {
    const delta = Number((ageFoot.series.female[i] - ageFoot.series.male[i]).toFixed(2));
    const stored = ageFoot.differenceFemaleMinusMale[i];
    if (Math.abs(delta - stored) > 0.011) {
      fail("age-foot-length difference mismatch at " + age + ": " + stored + " vs " + delta);
    }
  });
}

const bmi = readJson("bmi-foot-plumpness.json");
if (bmi) {
  if (!sameLength(bmi.bmi, bmi.series.female, bmi.series.male)) {
    fail("bmi-foot-plumpness series lengths differ");
  } else {
    ok("bmi-foot-plumpness aligned length=" + bmi.bmi.length);
  }
}

const radar = readJson("age-radar-profiles.json");
if (radar) {
  radar.series.forEach(function (row) {
    if (row.values.length !== radar.indicators.length) {
      fail("radar age " + row.age + " value count mismatch");
    }
    if (row.values.some(function (v) { return v < 0 || v > 100; })) {
      fail("radar age " + row.age + " has values outside 0–100");
    }
  });
  ok("radar profiles " + radar.series.length + " ages");
}

const pressure = readJson("plantar-pressure.json");
if (pressure) {
  if (!sameLength(pressure.sites, pressure.series.typicalAdult, pressure.series.diabeticFoot)) {
    fail("plantar-pressure series lengths differ");
  } else {
    ok("plantar-pressure sites=" + pressure.sites.length);
  }
}

const thickness = readJson("tissue-thickness.json");
if (thickness) {
  const n = thickness.siteCount;
  ["typicalAdultMm", "diabeticFootMm", "differenceUm"].forEach(function (key) {
    if (!isNumberArray(thickness.series[key]) || thickness.series[key].length !== n) {
      fail("tissue-thickness " + key + " expected " + n + " numbers");
    }
  });
  ok("tissue-thickness sites=" + n);
}

const symmetry = readJson("foot-symmetry.json");
if (symmetry) {
  symmetry.series.forEach(function (row) {
    const c = row.counts;
    const sumParts = c.same + c.leftLarger10to20 + c.leftLargerOver20 + c.rightLargerOver20 + c.rightLarger10to20;
    if (sumParts !== row.total) {
      fail("foot-symmetry age " + row.age + " total mismatch");
    }
  });
  ok("foot-symmetry ages=" + symmetry.series.length);
}

if (errors.length) {
  console.error("DATA VALIDATION FAILED");
  errors.forEach(function (e) { console.error(" - " + e); });
  process.exit(1);
}

console.log("DATA VALIDATION OK");
info.forEach(function (m) { console.log(" - " + m); });

File: scripts/validate_data.py
#!/usr/bin/env python3
"""Sanity-check extracted personal datasets before publishing examples."""

from __future__ import annotations

import json
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
DATA = ROOT / "data"


def fail(message: str) -> None:
    print(f"FAIL: {message}", file=sys.stderr)
    raise SystemExit(1)


def load(name: str) -> dict:
    path = DATA / name
    if not path.exists():
        fail(f"missing {path}")
    return json.loads(path.read_text(encoding="utf-8"))


def expect(cond: bool, message: str) -> None:
    if not cond:
        fail(message)


def main() -> None:
    catalog = load("catalog.json")
    for name in catalog["datasets"]:
        load(name)

    height = load("height-shoe-size.json")
    female = height["series"]["female"]["points"]
    male = height["series"]["male"]["points"]
    expect(len(female) == 199, f"female points {len(female)} != 199")
    expect(len(male) == 199, f"male points {len(male)} != 199")
    for sex, points in (("female", female), ("male", male)):
        for i, pair in enumerate(points):
            expect(len(pair) == 2, f"{sex}[{i}] is not a pair")
            expect(70 < pair[0] < 170, f"{sex}[{i}] height {pair[0]} out of range")
            expect(8 < pair[1] < 30, f"{sex}[{i}] shoe last {pair[1]} out of range")
    expect(height["series"]["female"]["stats"]["pearson_r"] > 0.8, "female correlation too weak")
    expect(height["series"]["male"]["stats"]["pearson_r"] > 0.8, "male correlation too weak")

    age = load("age-foot-length.json")
    expect(len(age["ages"]) == 9, "expected 9 ages")
    expect(len(age["female_cm"]) == 9, "female length series length")
    expect(len(age["male_cm"]) == 9, "male length series length")
    for i, (g, b, d) in enumerate(zip(age["female_cm"], age["male_cm"], age["female_minus_male_cm"])):
        expect(abs((g - b) - d) < 1e-9, f"age diff mismatch at index {i}")

    bmi = load("bmi-foot-ratio.json")
    expect(len(bmi["bmi"]) == 13, "bmi bins")
    expect(len(bmi["female_ratio"]) == 13, "female ratio")
    expect(0 < bmi["ring_charts"]["thin_feet"]["highlighted_share"] < 1, "thin share")
    expect(0 < bmi["ring_charts"]["plump_feet"]["highlighted_share"] < 1, "plump share")

    radar = load("growth-radar.json")
    expect(len(radar["axes"]) == 7, "radar axes")
    expect(radar["series"]["12岁"][4] == 71, "12-year shoe last should be 77-6")
    expect(radar["series"]["12岁"][5] == 71, "12-year foot length should be 79-8")

    bilateral = load("bilateral-symmetry.json")
    expect(len(bilateral["points"]) == 13, "ages 2-14")
    expect(bilateral["points"][0]["age"] == "2岁", "first age label")

    plantar = load("plantar-pressure.json")
    expect(len(plantar["sites"]) == 7, "pressure sites")
    expect(len(plantar["normal_pa"]) == 7, "normal pressure")
    expect(len(plantar["diabetic_foot_pa"]) == 7, "diabetic pressure")

    thickness = load("tissue-thickness.json")
    expect(len(thickness["normal_adult_mm"]) == 14, "thickness sites")
    expect(len(thickness["diabetic_foot_adult_mm"]) == 14, "diabetic thickness")
    expect(len(thickness["difference_labelled_um"]) == 14, "um series")

    print("ok: 7 datasets passed integrity checks")


if __name__ == "__main__":
    main()
